boring/src/ssl/mod.rs

```diff
@@ -1707,6 +1732,33 @@ impl SslContextBuilder {
         }
     }
 
+    fn set_safe_defaults(&mut self) -> Result<(), ErrorStack> {
+        // Choose a curves list suitable for the given feature set.
+        let curves = if cfg!(feature = "kx-pq-preferred") {
+            if cfg!(feature = "kx-fips-required") {
+                Some("P256Kyber768Draft00:P-256:P-384")
+            } else {
+                Some("X25519Kyber768Draft00:X25519:P256Kyber768Draft00:P-256:P-384")
+            }
+        } else if cfg!(feature = "kx-pq-supported") {
+            if cfg!(feature = "kx-fips-required") {
+                Some("P-256:P-384:P256Kyber768Draft00")
```

Comment:
This is fine if it's client-side, but not if it's server-side. For server-side we want to have PQ in the front.

I think this sets the defaults for both client and server side?

Reply:
Many services act as both client and server, so the current approach will not work.

Reply:
The problem is that the client and server need to have different default curve preferences. Options:
1. Remove `kx-fips-required` flag and force clients to call `SslContextBuilder::set_curves_list()` manually.
2. Wait to resolve default preferences until `Ssl::accept()` or `Ssl::connect()` is called. Background: The `SslContext` does not know whether it will be used for the client or server. This is partially determined by `SslMethod`, but there are methods that do not specify the role. What we could do instead is propagate a flag from `SslContextBuilder` to `SslContext` to `Ssl` that indicates whether the default curve preferences have been overwritten by the user, and if not, set the curves list just before `accept` or `connect`.

The second option is preferable because it requires no intervention on the part of the user.

Reply:
Option 2 is pretty tricky as it turns out! I considered a few options for adding the flag:

2(a) Wrap `SslContext`, i.e., `pub struct WrappedSslContext{ inner: SslContext, curve_preference_set_by_user: bool }`. This is the safest option because we don't have to mess with the FFI bits. However, this requires a major refactor (most consumers of `SslContext` expect a thin wrapper around `SSL_CTX`).

2(b) Change the type wrapped by `SslContext` define a C struct that contains the `SSL_CTX` and the flag. This is not ideal because it requires doing surgery on `unsafe` code.

3(b) Create a macro similar to `generic_foreign_type_and_impl_send_sync!` that wraps a foreign type _and_ and a rust struct, perhaps one that is `Default` and `Copy` to keep things simple. The details are a bit tricky, but this seems promising and I'm still working on it.

A couple of other options I wanted to flag, which would be simpler to implement:

(4) Patch boringSSL with a method that checks if the user changed the default curve preferences. This is not ideal because the implementation would subject to change as boringSSL evolves.

(5) When the feature flags are set **disable** manually setting curve preferences, i.e., set_curve_list and set_curves would return an error if any of "kx-pq-preferred", "kx-fips-required", etc. are used. This is easy to implement but is more restrictive than we'd like.

Reply:
I opted for (5) due to its simplicity and fail safety: https://github.com/cloudflare/boring/pull/145/commits/9eedf6178afda50703f4e1e1b07a15b76586a68e

Reply:
I think (5) is elegantly pragmatic. If we have a user that needs to set curve preferences, we can change to something like (4): IIRC BoringSSL uses an empty list if curve preferences aren't set.